### PR TITLE
ARROW-5631: [C++] Fix FindBoost targets with cmake3.2

### PR DESCRIPTION
**Pull request by fsaintjacques**
cmake's FindBoost pre 3.5 did not export IMPORTED targets. This also fixes the newly introduced trusty docker image.

### CODE REVIEW

File: cpp/cmake_modules/ThirdpartyToolchain.cmake
Comment (fsaintjacques) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake]: I'm not sure why this was like this, if someones wants to clarify, I'll gladly add a comment to the PR (@xhochy, @wesm).

**Comment by fsaintjacques** [comment]
A note that this patch does not fully fix the cmake3.2 issue, see https://jira.apache.org/jira/browse/ARROW-5631?filter=-1 

But at least it gets a cmake build directory populated.

**Comment by wesm** [comment]
Looks like @kou introduced that logic here

https://github.com/apache/arrow/commit/f014d76c68a9a977a19f87f7ea77511e4753bb8c

I'm not sure what it's for as we don't use this flag other places in our build system. Some projects use the flag to globally toggle between building SHARED/STATIC when using `add_library` but we are explicit about this everywhere.

**Comment by fsaintjacques** [comment]
@kou afaik, find_package shouldn't be affected by BUILD_SHARED_LIBS?

**Comment by wesm** [comment]
I think it's fine to keep the stashing; it seems like it was more likely future proofing

**Comment by fsaintjacques** [comment]
Don't merge for now, it doesn't fix the docker cmake32 image.

File: cpp/cmake_modules/ThirdpartyToolchain.cmake
Comment (kou) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake]: This is needed to work with `BoostConfig.cmake` provided by Boost (not CMake).
The Boost package provided by MSYS2 installed `BoostConfig.cmake` at the time.
But the current Boost package doesn't install `BoostConfig.cmake`: https://github.com/msys2/MINGW-packages/pull/5326/files

See also: https://github.com/msys2/MINGW-packages/issues/5233

There is no platform that needs this. So we can remove this for now but we may need this again in the future.

**Comment by fsaintjacques** [comment]
rebase with master, ready to land (once it passes tests). This passes locally, minus the linking error fixed by https://github.com/apache/arrow/pull/4635.

**Comment by kou** [comment]
It seems that the current changes don't fix anything.
Should we really fix this problem or update title and description of this pull request?

**Comment by fsaintjacques** [comment]
@kou previously cmake didn't even succeed at generating a build folder. The ctest failures are fixed in the PR linked in previous comment (#4635).

**Comment by fsaintjacques** [comment]
@ursabot crossbox --help

**Comment by fsaintjacques** [comment]
@ursabot crossbow --help

**Comment by fsaintjacques** [comment]
@ursabot crossbow test

**Comment by fsaintjacques** [comment]
@ursabot crossbow test -g docker

**Comment by fsaintjacques** [comment]
@ursabot crossbow test docker docker-cpp docker-cpp-cmake32 docker-rust

**Comment by fsaintjacques** [comment]
@ursabot crossbow test docker-cpp docker-cpp-cmake32 docker-rust

**Comment by fsaintjacques** [comment]
@ursabot crossbow test docker-c_glib

**Comment by fsaintjacques** [comment]
@kszucs do you know why the badge is at failed? The last one are waiting to be run.

**Comment by fsaintjacques** [comment]
@kou could you check locally with this patch why c_glib docker image fails? I've exausted my debugging of  meson. I think that you need to namespace gandiva's `enums.c`.

```
gandiva-glib/meson.build:50:0: ERROR:  Tried to create target "enums.c", but a target of that name already exists.
```

I'll let you decide if renaming breaks anything.

**Comment by fsaintjacques** [comment]
@ursabot crossbow test docker-c_glib

**Comment by fsaintjacques** [comment]
@ursabot crossbow test docker-r

**Comment by fsaintjacques** [comment]
@ursabot crossbow test -g docker

**Comment by fsaintjacques** [comment]
This fixes the most pressing issues, notably
- docker-cpp
- docker-cmake32

I fixed also fixed docker-rust, and partly advanced on docker-r and docker-c_glib.

**Comment by fsaintjacques** [comment]
@ursabot crossbow test docker-cpp docker-cpp-cmake32

File: c_glib/Dockerfile
Comment (kou) [inline c_glib/Dockerfile]: Is this needed?
The added path is included in system `pkg-config` path. So it'll be searched without this.

**Review by kou — COMMENTED** [review]
> ```text
> gandiva-glib/meson.build:50:0: ERROR:  Tried to create target "enums.c", but a target of that name already exists.
> ```

This is a Meson bug. It has been fixed in the latest Meson. But the latest Meson (0.51.0) has a environment variable related bug: https://github.com/mesonbuild/meson/issues/5503 https://github.com/mesonbuild/meson/issues/5502

So we need to use 0.50.1 for now.

**Comment by kou** [comment]
BTW, I couldn't find any build log on CircleCI.

https://circleci.com/gh/ursa-labs/crossbow/tree/ursabot-46-circle-docker-c_glib shows

image

And "build" link shows

image

File: c_glib/Dockerfile
Comment (wesm) [reply c_glib/Dockerfile]: I'm going to revert this for now

File: r/Dockerfile
Comment (wesm) [reply r/Dockerfile]: fixing

**Review by wesm — APPROVED** [review]
+1. Made a couple of tweaks. Merging -- please open follow up JIRAs about other docker failures

**Pull request closed without merging**
